openra/services/screenshot_repository.py

```diff
@@ -0,0 +1,93 @@
+import io
+from dependency_injector.wiring import Provide, inject
+from django.conf import os
+from django.contrib.auth.models import User
+from django.core.files.uploadedfile import UploadedFile
+from django.utils import timezone
+from fs.base import FS, copy
+from openra.classes.exceptions import ExceptionBase
+from openra.classes.file_location import FileLocation
+from openra.classes.screenshot_resource import ScreenshotResource
+from openra.models import Screenshots
+from PIL import Image
+
+from openra.services.uploaded_file_importer import UploadedFileImporter
+
+
+class ScreenshotRepository:
+
+    _data_fs: FS
+    _uploaded_file_importer: UploadedFileImporter
+
+    @inject
+    def __init__(
+        self,
+        data_fs: FS = Provide['data_fs'],
+        uploaded_file_importer: UploadedFileImporter = Provide['uploaded_file_importer']
+    ):
+        self._data_fs = data_fs
+        self._uploaded_file_importer = uploaded_file_importer
+
+    def create_from_uploaded_file(self, uploaded_file: UploadedFile, user: User, resource: ScreenshotResource, map_preview: bool):
+
+        if uploaded_file.content_type not in ['image/jpeg', 'image/png', 'image/gif']:
+            raise ExceptionInvalidMimeType(uploaded_file.name, uploaded_file.content_type)
+
+        extension = uploaded_file.content_type.split('/')[1]
+
+        uploaded = self._uploaded_file_importer.import_file(
+            uploaded_file,
+            uploaded_file.name
+        )
+
+        image = Image.open(
+            uploaded.get_file_clone()
```

Comment:
Does this need to be a clone? Why not use the original?

Reply:
It's in a file system - in this case a memory one so might not necessarily have a stored location on the disk. 

The Image class needed an actual image path or a ByteIO object, so in order to satisfy that, it needed to clone the data from the file system into a BytesIO object.

Reply:
The main reason for it being on a pseudo file system is do that it can be used, transferred or whatever in a consistent manner throughout the code, without it needing to know about where it's stored.

This means that if all internal classes use this format, there is less likely to need to be code dealing with the locations where the file is only in temporary locations anyway.

It also means that the file in this case is just in memory, so should be really fast to read and process etc. Instead of needing to cause disk IO - although I guess django *might* already use the disk for the initial upload.

Reply:
Without going overboard on this explanation - there's also the benefit, that where it's in it's own memory file system, there's no chance of file path collisions or anything like that either.